Signal/src/util/OWSBackupExportJob.m

```diff
@@ -0,0 +1,640 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+#import "OWSBackupExportJob.h"
+#import "Signal-Swift.h"
+#import <SignalServiceKit/NSData+Base64.h>
+#import <SignalServiceKit/NSDate+OWS.h>
+#import <SignalServiceKit/OWSBackgroundTask.h>
+#import <SignalServiceKit/OWSBackupStorage.h>
+#import <SignalServiceKit/OWSError.h>
+#import <SignalServiceKit/OWSFileSystem.h>
+#import <SignalServiceKit/TSAttachmentStream.h>
+#import <SignalServiceKit/TSMessage.h>
+#import <SignalServiceKit/TSThread.h>
+#import <SignalServiceKit/Threading.h>
+#import <SignalServiceKit/YapDatabaseConnection+OWS.h>
+#import <YapDatabase/YapDatabase.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+NSString *const kOWSBackup_ExportDatabaseKeySpec = @"kOWSBackup_ExportDatabaseKeySpec";
+
+@interface OWSAttachmentExport : NSObject
+
+@property (nonatomic, weak) id<OWSBackupJobDelegate> delegate;
+@property (nonatomic) NSString *jobTempDirPath;
+@property (nonatomic) NSString *attachmentId;
+@property (nonatomic) NSString *attachmentFilePath;
+@property (nonatomic, nullable) NSString *tempFilePath;
+@property (nonatomic, nullable) NSString *relativeFilePath;
+
+@end
+
+#pragma mark -
+
+@implementation OWSAttachmentExport
+
+- (void)dealloc
+{
+    // Surface memory leaks by logging the deallocation.
+    DDLogVerbose(@"Dealloc: %@", self.class);
+
+    // Delete temporary file ASAP.
+    if (self.tempFilePath) {
+        [OWSFileSystem deleteFileIfExists:self.tempFilePath];
+    }
+}
+
+// On success, tempFilePath will be non-nil.
+- (void)prepareForUpload
+{
+    OWSAssert(self.jobTempDirPath.length > 0);
+    OWSAssert(self.attachmentId.length > 0);
+    OWSAssert(self.attachmentFilePath.length > 0);
+
+    NSString *attachmentsDirPath = [TSAttachmentStream attachmentsFolder];
+    if (![self.attachmentFilePath hasPrefix:attachmentsDirPath]) {
+        DDLogError(@"%@ attachment has unexpected path.", self.logTag);
+        OWSFail(@"%@ attachment has unexpected path: %@", self.logTag, self.attachmentFilePath);
+        return;
+    }
+    NSString *relativeFilePath = [self.attachmentFilePath substringFromIndex:attachmentsDirPath.length];
```

Comment:
Could we simplify with:

`NSString *relativeFilePath = self.attachmentFilePath.lastPathComponent`

Reply:
No, because that assumes that the relative paths only have a single component.  Attachments all used to live in the same "attachment directory", but as of 6 months ago, we modified them to live in subdirectories so that they could use their "source filename" (which is used when sharing them with the share sheet) without fear of multiple attachments with the same filename conflicting with each other.